### PR TITLE
User-LD package

### PR DESCRIPTION
**Pull request by tanmoy7989**
**Summary**

This package provides the first implementation (see references) of the Local Density (LD) potential  that comes from the Shell group in Chemical Engineering at UC Santa Barbara. Although, essentially it is only a single pair.h and pair.cpp file, other research groups have since implemented variants of their own LD potentials, and we hope they will be able to contribute by simply adding to this package. This is why we decided to make it a standalone package.

**Author(s)**

Tanmoy Sanyal
Chemical Engineering, UC Santa Barbara
currently at Bioengineering, UC San Francisco
email: tanmoy dot 7989 at gmail.com

David Rosenberger
Computational Physical Chemistry, TU Darmstadt
email: rosenberger at cpc.tu-darmstadt.de

Prof. M. Scott Shell
Chemical Engineering, UC Santa Barbara
email: shell at engineering.ucsb.edu

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

Backwards compatible

**Implementation Notes**

The code draws heavily from the pair_eam styles. For further information please read the manual pdf. Two example systems have been provided and tested in both serial and mpi modes.

**Post Submission Checklist**

_Please check the fields below as they are completed **after** the pull request has been submitted. Delete lines that don't apply_

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] The added/updated documentation is integrated and tested with the documentation build system
- [x] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [x] A package specific README file has been included or updated
- [x] One or more example input decks are included

**Further Information, Files, and Links**

References
1) [Sanyal and Shell, Coarse-grained models using local density potentials optimized with the relative entropy: Application to implicit solvation}, Journal of Chemical Physics, 2016, 145 (3), 034109](https://aip.scitation.org/doi/abs/10.1063/1.4958629)
 
2) [Sanyal and Shell, Transferable coarse-grained models of liquid-liquid equilibrium using local density potentials optimized with the relative entropy, Journal of Physical Chemistry B, 122 (21), 5678-5693](https://pubs.acs.org/doi/abs/10.1021/acs.jpcb.7b12446)

3) [Rosenberger, Sanyal, Shell and van der Vegt, Transferability of local density assisted implicit solvation models for homogenenous fluid mixtures, Journal of Chemical Physics, 2019, 151 (4), 044111](https://pubs.acs.org/doi/abs/10.1021/acs.jctc.8b01170)

### CODE REVIEW

**Comment by akohlmey** [comment]
> This package provides the first implementation (see references) of the Local Density (LD) potential that comes from the Shell group in Chemical Engineering at UC Santa Barbara. Although, essentially it is only a single pair.h and pair.cpp file, other research groups have since implemented variants of their own LD potentials, and we hope they will be able to contribute by simply adding to this package. This is why we decided to make it a standalone package.

Sorry, but we can discuss moving this into a package of its own when there are actually multiple additional related pair styles submitted and a creation of an independent package thus justified. For now it would be better to submit it as a contribution to the USER-MISC package. At that point also the name of the package would have to be different as USER-LD is too generic. Making this a contribution to USER-MISC requires changes to the documentation and moving the examples to a different folder. I have some more comments and requests for changes, that I will annotate to individual files.

File: src/USER-LD/README [hunk 1]
```diff
@@ -0,0 +1,7 @@
+USER-LD package
```
Comment (akohlmey) [inline src/USER-LD/README]: This file would not be needed for contributing the pair style to USER-MISC, but a corresponding entry to the README file in that folder should be added.

File: src/USER-LD/examples/benzene_water/benzene_water.data [hunk 2]
```diff
@@ -0,0 +1,1406 @@
+LAMMPS data file for 380 CG benzene, 1000 CG water particles
```
Comment (akohlmey) [inline src/USER-LD/examples/benzene_water/benzene_water.data]: Examples files should never be in the source tree. For a USER-MISC package, the corresponding folder would be `examples/USER/misc/<name>`, e.g. using `local_density` for `<name>`

File: src/USER-LD/manual/ld_manual.tex [hunk 3]
```diff
@@ -0,0 +1,149 @@
+\documentclass[12pt,a4paper]{article}
```
Comment (akohlmey) [inline src/USER-LD/manual/ld_manual.tex]: Documentation should not be in the source tree. Rather this document should be converted into a file `doc/src/pair_local_density.txt` in a style equivalent to existing files in that folder, plus possibly a file Howto_local_density.txt with additional information, e.g. how to build a complete input deck and its support files.
For latex style typesetting, MathJaX can be used. We plan to convert the documentation to using RST soon (currently there is a tool for conversion) and then will support LaTeX typesetting in all versions of the manual. All explicit image files from `doc/src/Eqs` will be converted to MathJaX in the process, so you can save us some time and effort by doing this directly.

File: src/USER-LD/pair_localdensity.h [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+-------------------------------------------------------------------------
+   pair_localdensity written by:
+   Tanmoy Sanyal and M. Scott Shell from UC Santa Barbara
+   David Rosenberger: TU Darmstadt
+-------------------------------------------------------------------------*/   
+    
+
+#ifdef PAIR_CLASS
+
+PairStyle(localdensity,PairLOCALDENSITY)
```
Comment (akohlmey) [inline src/USER-LD/pair_localdensity.h]: for improved consistency with the rest of LAMMPS, please name the pair style `local/density`, the class `PairLocalDensity` and source and documentation files `pair_local_density.<ext>`

File: tools/reorderLammpsREMD.py [hunk 5]
```diff
@@ -0,0 +1,567 @@
+#!/usr/bin/env python
```
Comment (akohlmey) [inline tools/reorderLammpsREMD.py]: Please don't include code from a different pull request.

File: src/USER-LD/pair_localdensity.cpp [hunk 6]
```diff
@@ -0,0 +1,883 @@
...
+     spline     matrix that collects b,c,d
+     
+	 
+	 other parameters:
+          fpa		derivative of function at x=a
+          fpb		derivative of function at x=b
+*/
+                     
+     double *dl, *dd, *du;
+     double *b, *c, *d;
```
Comment (akohlmey) [inline src/USER-LD/pair_localdensity.cpp]: the pointer variable `b` here shadows the class member pointer variable `b`. One of them should be renamed.

File: src/USER-LD/pair_localdensity.cpp [hunk 7]
```diff
@@ -0,0 +1,883 @@
...
+  memory->destroy(c4);
+  memory->destroy(c6);
+  memory->destroy(uppercut);
+  memory->destroy(lowercut);
+  memory->destroy(uppercutsq);
+  memory->destroy(lowercutsq);
+  memory->destroy(frho);
+  memory->destroy(rho);
+
+  delete [] a;
```
Comment (akohlmey) [inline src/USER-LD/pair_localdensity.cpp]: `a` and `b` are allocated via `memory->create()` and not `new`, thus they should be freed using `memory->destroy()`. The memory class uses `malloc()/free()` internally which should not be mixed with `new/delete`. Some compilers will throw errors, if this mix happens.

File: src/USER-LD/README (hunk 1 above)
Comment (tanmoy7989) [reply src/USER-LD/README]: done

**Pull request closed without merging**